src/app/components/table/table.component.ts

```diff
@@ -27,26 +33,46 @@ export class TableComponent implements OnInit {
 
   constructor(private readonly mainService: MainService) {}
 
-  ngOnInit() {
-    (async () => {
-      await (
-        await this.mainService.getAllElementTypes().toPromise()
-      ).forEach((type) => {
-        const _ =
-          this.elementTypes.find((et) => et.uri === type.uri.split('@').shift()) ||
-          this.elementTypes.push({
-            ...type,
-            uri: type.uri.split('@').shift()
-          });
-      });
-    })();
+  ngOnInit(): void {
+    this.initElemTypes();
+  }
+
+  ngOnDestroy(): void {
+    this.toDestroy$.next();
+    this.toDestroy$.complete();
+  }
+
+  /**
+   * split input text by @ char and returns first elem if found.
+   * @param text input text to split by `@`
+   */
+  private readonly takeFirstUriWord = (text): string => (text || '').split('@').shift();
+
+  private initElemTypes = (): void => {
+    this.mainService.getAllElementTypes()
+      .pipe(
+        map(types => types.reduce((accum, el) => {
+          const elUriWord = this.takeFirstUriWord(el.uri);
+          // don't push duplicate elem
+          if (!accum.some(t => this.takeFirstUriWord(t.uri) === elUriWord)) {
+            accum.push({...el, uri: elUriWord});
+          }
+          return accum;
+          }, [] as ElementType[])),
+        takeUntil(this.toDestroy$)
+      ).subscribe({
+      next: types => this.elementTypes = types,
```

Comment:
So this component got some complexity code which was doing some transformation based on local requirement, I move this to a separate function so it'll be easier to debug and capture the error following by the `observer` pattern ensuring cancellation of process and mem leak during component destroy hook.

Like if all transformations are done inside `IIFE` with the nested `async, await` we still can use `.then` or `.catch` after `.toPromise()` but as far as the term of readable,testable, reusable code goes, probably don't pollute the life-cycle-hooks and make a chunk of function which can perform a micro task where everyone can get started with less complexity. 

I might be inaccurate in many things I have mentioned but that what's came up in my mind during refactorization.